--- src/xalanc/Utils/XalanMsgLib/Makefile	2017-09-04 14:13:20.589053722 -0400
+++ src/xalanc/Utils/XalanMsgLib/Makefile	2017-09-01 10:44:56.290642979 -0400
@@ -17,7 +17,7 @@
 # Output directory
 OUTDIR := ../../../..
 
-INSTALL = /usr/bin/install -c
+INSTALL = /bin/install -c
 INSTALL_PROGRAM = ${INSTALL}
 INSTALL_DATA = ${INSTALL} -m 644
 INSTALL_SCRIPT = ${INSTALL}
@@ -32,7 +32,9 @@
 .PHONY: doClean install
 
 $(MESSAGELIB) : ${XSL_OBJ_DIR}/XalanMsgLib.o 
-	$(MAKE_SHARED_LOC) $(EXTRA_LOC_LINK_OPTIONS) $^ -o $@
+	#$(MAKE_SHARED_LOC) $(EXTRA_LOC_LINK_OPTIONS) $^ -o $@
+	ar -cq $@ $^   
+	cp $@ ${XSL_LIB_DIR}/libxalanMsg.a
 ifneq ($(PLATFORM), OS390)
 	rm -f $(XSL_LIB_DIR)/$(LOC_LIBNAME)$(SHLIBSUFFIX) && \
 		$(LN) -fs $(LOC_LIB) $(XSL_LIB_DIR)/$(LOC_LIBNAME)$(SHLIBSUFFIX)
