--- pwiz/utility/misc/Stream.hpp	2017-10-05 03:05:17.000000000 -0400
+++ pwiz/utility/misc/Stream.hpp	2017-10-06 11:22:17.716789600 -0400
@@ -27,12 +27,12 @@
 #include <iomanip>
 #include <boost/iostreams/operations.hpp>
 #include "pwiz/utility/misc/optimized_lexical_cast.hpp"
-#include <boost/nowide/fstream.hpp>
-#include <boost/nowide/iostream.hpp>
-#include <boost/nowide/args.hpp>
+//#include <boost/nowide/fstream.hpp>
+//#include <boost/nowide/iostream.hpp>
+//#include <boost/nowide/args.hpp>
 
 namespace bio = boost::iostreams;
-namespace bnw = boost::nowide;
+//namespace std = boost::nowide;
 
 using std::ios;
 using std::iostream;
@@ -42,9 +42,9 @@
 using std::istream_iterator;
 using std::ostream_iterator;
 
-using bnw::fstream;
-using bnw::ifstream;
-using bnw::ofstream;
+using std::fstream;
+using std::ifstream;
+using std::ofstream;
 
 using std::stringstream;
 using std::istringstream;
@@ -56,9 +56,9 @@
 using std::streamoff;
 using std::streamsize;
 
-using bnw::cin;
-using bnw::cout;
-using bnw::cerr;
+using std::cin;
+using std::cout;
+using std::cerr;
 using std::endl;
 using std::flush;
 
--- pwiz/utility/misc/random_access_compressed_ifstream.cpp	2017-10-05 03:05:17.000000000 -0400
+++ pwiz/utility/misc/random_access_compressed_ifstream.cpp	2017-10-06 11:35:18.469883500 -0400
@@ -82,7 +82,7 @@
 #include "pwiz/utility/misc/Std.hpp"
 #include <boost/thread.hpp>
 #include <boost/filesystem/path.hpp>
-#include <boost/filesystem/detail/utf8_codecvt_facet.hpp>
+//#include <boost/filesystem/detail/utf8_codecvt_facet.hpp>
 
 
 namespace pwiz {
@@ -1047,8 +1047,8 @@
         return false;
     }
 
-    boost::filesystem::detail::utf8_codecvt_facet utf8;
-    this->handle = new boost::iostreams::file_descriptor_source(boost::filesystem::path(path, utf8));
+    //boost::filesystem::detail::utf8_codecvt_facet utf8;
+    this->handle = new boost::iostreams::file_descriptor_source(boost::filesystem::path(path));
     this->flen = 0;
     this->desired_readbuf_len = 0;
     // dynamic read buffer sizing - start small
--- pwiz/utility/misc/SHA1.cpp	2017-10-05 03:05:17.000000000 -0400
+++ pwiz/utility/misc/SHA1.cpp	2017-10-06 11:37:07.910771500 -0400
@@ -16,7 +16,7 @@
 #include "SHA1.h"
 #include <boost/nowide/fstream.hpp>
 #include <boost/filesystem/operations.hpp>
-#include <boost/filesystem/detail/utf8_codecvt_facet.hpp>
+//#include <boost/filesystem/detail/utf8_codecvt_facet.hpp>
 #include <boost/interprocess/file_mapping.hpp>
 #include <boost/interprocess/mapped_region.hpp>
 
@@ -193,7 +193,7 @@
 		ifstream fpIn(tszFileName, std::ios::binary);
 		if (!fpIn) return false;
 
-		const INT_64 lFileSize = boost::filesystem::file_size(boost::filesystem::path(tszFileName, boost::filesystem::detail::utf8_codecvt_facet()));
+		const INT_64 lFileSize = boost::filesystem::file_size(boost::filesystem::path(tszFileName));
 		const INT_64 lMaxBuf = SHA1_MAX_FILE_BUFFER;
 		char vData[SHA1_MAX_FILE_BUFFER];
 		INT_64 lRemaining = lFileSize;
